Implement Ctrl-o to execute a line from history and jump to the next entry

hupfdule opened the pull request:
This commit replicates bashs binding to Ctrl-o.

Ctrl-o exists in bash to execute a command from the history (as Enter
does), but then display the next entry from the history. Another press
of Ctrl-o then executes that next line and displays the entry after it.

For example if the are the following entries in the history:

  echo 1
  echo 2
  echo 3

and the entry "echo 1" is the currently selected entry in re-search,
Ctrl-o would execute "echo 1" and then display "echo 2". Another press
of Ctrl-o would display "echo 3" and so on.

This allows repeating the same sequence of commands in history just by
pressing Ctrl-o multiple times.

To allow for displaying the result of the executed commands while still
inside re-search, the corresponding fish function needed to be changed
to specify the temporary file containing the line to execute as an
argument to re-search instead of always printing to stdout and using
redirection.

This PR depends on #20.

hupfdule committed:
Implement Ctrl-o to execute a line from history and jump to the next entry

This commit replicates bashs binding to Ctrl-o.

Ctrl-o exists in bash to execute a command from the history (as Enter
does), but then display the next entry from the history. Another press
of Ctrl-o then executes that next line and displays the entry after it.

For example if the are the following entries in the history:

  echo 1
  echo 2
  echo 3

and the entry "echo 1" is the currently selected entry in re-search,
Ctrl-o would execute "echo 1" and then display "echo 2". Another press
of Ctrl-o would display "echo 3" and so on.

This allows repeating the same sequence of commands in history just by
pressing Ctrl-o multiple times.

To allow for displaying the result of the executed commands while still
inside re-search, the corresponding fish function needed to be changed
to specify the temporary file containing the line to execute as an
argument to re-search instead of always printing to stdout and using
redirection.

hupfdule committed:
Display error if append to history didn't work.

jbonjean commented:
I didn't know about this bash feature, very interresting. But I have mixed feelings about the exec part. I'm not sure that it's the good thing to do in every situation and feel it's a bit outside the scope of re-search. Let me think about it.

hupfdule commented:
My first idea was to return a certain exit code and handle that in the re_search.fish function.
But I didn't find a way to do so. In any case it looked quite complicated. That brought me to the idea to handle this in re-search internally. This makes it less complicated, actually and after playing around with it for a while I really like it even prefer to stay inside re-search for this operation.

hupfdule committed:
Removed duplicate append_to_history() method

append_to_history() was also introduced with commit
c224a91976ea8817c9e6e4c47c43ca7d52c19f73 and has a better implementation
as it does not error out if the maximum history size was reached.

hupfdule committed:
Do not reduce history on Ctrl-q

This commit changes the behaviour of Ctrl-q in that it does not remove
non-matching history entries, but instead just stores the subsearch
terms on compares them additionally to the current search term on
searching.

This is necessary to support resetting of subsearches (drop the last
subsearch) and also play nicely with the new support for Ctrl-o (as a
press of Ctrl-o should jump the next entry in the whole history, not the
next entry matching the subsearches).

Theoretically that reduces the performance of searching a bit for each
additional subsearch (as more comparisons must be done for each history
entry) whereas the previous implementation should have gotten faster for
each additional subsearch (as the number of history entries was reduced
with each subearch). However, no measurement has been done to verify
that this is actually a noticable delay.

hupfdule committed:
Drop last subsearch term with Ctrl-x

hupfdule committed:
Only search forward if action == SEARCH_FORWARD

Since we now have new actions SCROLL and EXECUTE we need to check this
explicitly.

hupfdule committed:
Recalculate the the current search index if necessary

If we start a search from the middle of the history (after scroll events
via Ctrl-p/Ctrl-n) we need to calculate the number of matching history
entries before the current index.

jbonjean commented:
OK, so I'm not going to make you wait any longer, I'm not going to accept this PR. I'm not convinced this won't interfere with other shell features. Once again, I would accept it if it was disabled by default (`#ifdef`), but I could understand if you prefer to just maintain this in your fork.

hupfdule commented:
Ok, thanks for fhe info.

I'd like to leave this PR open until I have decided on whether I want to wrap it inside `#ifdef` or just maintain my own fork if that is ok for you. 

Unfortunately I have less time now at the moment for working on that.

jbonjean commented:
No problem, we can keep this open.

hupfdule commented:
OK,

I have decided that it would be easier for me to just maintain my own fork.
I therefore close this issue now.

Thanks for your help!
